cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,266 @@
+  * However, if the biggest pod in the cluster was a database, we won't want to keep a hot standby node for the pod to move to
+* [R7] Possible to provide extra space even for every pod which is not biggest but do request some cpu(< max(pod_request_cpu)) and memory(< max(pod_request_mmory))
+  * Say, the "biggest pods" request 1000m cpu/1G mem, 10m cpu/10G mem respectively
+  * Even if they could fit within existing free space, a 500m(<1000m cpu)/5G mem pod has chances not to fit any node 
+* [R8] Reserve space even for the pods having anti pod affinity
+* [R9] Reserve space even for the pods having node selector/node affinity
+
+Version 3
+
+contains [R1]-[R9] and also what is below:
+
+* [R10] Works on considerably large clusters
+  * To move CA from beta to GA we need to handle 5000 nodes and 150 000 pods.
+ 
+### Specifications
+
+Version 1
+
+* [S1] TBD [R1-1]
+* [S2] Retain placeholder pods among iterations of CA and remove/re-assign/add only necessary placeholder pods. [R1-2]
+  * More concretely, for each CA iteration:
+  * Step 1: Calculate a `list of pushed-out placeholder pods`
+    * Iterate over existing placeholder pods to determine which one is "pushed out of the node" by actual pod(s)
+       * Iterate over ready nodes and see if `sum(pod requested cpu/mem)` exceeds a node's capacity. If it exceeds, push out placeholders until the sum becomes less than or equal to the node capacity.
+    * Disassociate the pushed-out placeholder pods with the formerly associated nodes
+  * Step 2: Calculate `desired extra capacity`
+  * Step 3: Calculate `desired number of placeholder pods` from `desired extra capacity`
+  * Step 4: Remove more placeholder pods if current extra capacity is larger than desired. Add more placeholder pods if desired extra capacity is increased.
+  * Step 5: Do usual scale-up/down
+  * Step 5: Retain the list of placeholder pods for the next iteration...
+* [S3] Every place holder pod should be small enough to fit within at least one of nodes [R2]
+  * Otherwise CA will produce no resource slack because it will be unable to add any node
+* [S4] Create exactly one placeholder replicaset which looks like managing all the placeholder pods [R3]
+* [S5] Provide `--extra-capacity-min-rate=<float>` to specify the rate of extra capacity compared to the total requested capacity [R4]
+
+Version 2
+
+contains [S1]-[S5] and also what is below:
+
+* [S6] Every placeholder pod should have the same or smaller size compared to the pod with most requested resource [R5]
+  * Otherwise CA will produce too much resource slack by adding an unnecessarily large node
+* [S7] Pods/Deploypments/ReplicaSets/Jobs marked via specific annotations are tried to pre-reserve space for them [R6]-[R9]
+  * Those pods are mirrored to equivalent placeholder pods, that can be bigger than standard placeholder pods, so that we can ensure the cluster to have extra space for them
```

Comment:
Let's say a user specified 10% slack as mentioned in the document. Some pods are resource governed (CPU/Memory), but others governed by _policy_ (say pod anti-affinity).

_What would it then mean to create a 10% slack when there are two completely different classes of constraints and disjoint properties governing resource estimation and allocation?_

The document as is focuses almost exclusively on the CPU/Memory angle. However my company uses mostly policy. In our _compute_ clusters we run 1 pod per node. We use pod anti-affinity to ensure each pod gets its own hardware. We don't use resource quotas (limits/requests) because all pods in such a cluster have the same workload characteristic, and have their needs automatically satisfied by being scheduled onto (identical) machines used in a node group. Our only requirement is to have dedicated machines of a certain size, so we can allocate a whole machine to every pod that is running (to avoid CPU cache misses etc), and that only one pod may run on a given node at any point in time.

I'm hoping anti-affinity is considered and prioritized for the _slack_ initiative, but the former is only mentioned in passing in this document. Some elaboration on how policy based constraints factor into the proposed design would be appreciated.